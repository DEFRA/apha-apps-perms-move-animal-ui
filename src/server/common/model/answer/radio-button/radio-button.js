import Joi from 'joi'
import { AnswerModel } from '../answer-model.js'
import { validateAnswerAgainstSchema } from '../validation.js'
import { NotImplementedError } from '../../../helpers/not-implemented-error.js'

/**
 * @param {RadioButtonConfig} config
 * @returns {Joi.Schema}
 */
const createRadioSchema = (config) => {
  const { emptyOptionText } = config.errors
  return Joi.object({
    [config.payloadKey]: Joi.string()
      .required()
      .valid(...Object.keys(config.options))
      .messages({
        'any.required': emptyOptionText,
        'any.valid': emptyOptionText,
        'string.empty': emptyOptionText,
        'any.only': emptyOptionText
      })
  })
}

/**
 * @typedef {{ label: string, hint?: string }} RadioOption
 * export @typedef {{
 *  payloadKey: string,
 *  options: Record<string, RadioOption>,
 *  errors: {
 *    emptyOptionText: string
 *  }
 * }} RadioButtonConfig
 */

/**
 * @template Payload
 * @augments AnswerModel<Payload>
 */
export class RadioButtonAnswer extends AnswerModel {
  /** @returns {RadioButtonConfig} */
  get config() {
    throw new NotImplementedError()
  }

  /** @returns {RadioButtonConfig} */
  static get config() {
    throw new NotImplementedError()
  }

  /**
   * @returns {Payload | undefined}
   */
  toState() {
    return this._data?.[this.config.payloadKey] ?? ''
  }

  get value() {
    return this._data?.[this.config.payloadKey]
  }

  get html() {
    const value = this._data?.[this.config.payloadKey]

    return this.config.options[value]?.label ?? ''
  }

  /**
   * @param {string | undefined} state
   * @returns {RadioButtonAnswer}
   */
  static fromState(state) {
    return new this(
      state !== undefined ? { [this.config.payloadKey]: state } : undefined
    )
  }

  validate() {
    return validateAnswerAgainstSchema(
      createRadioSchema(this.config),
      this._data ?? {}
    )
  }

  /**
   * @param {Payload} fields
   */
  _extractFields(fields) {
    return /** @type {Payload} */ ({
      [this.config.payloadKey]: fields[this.config.payloadKey]
    })
  }

  get viewModelItems() {
    const result = Object.entries(this.config.options).map(([key, value]) => ({
      id: key,
      value: key,
      text: value.label
    }))
    result[0].id = this.config.payloadKey
    return result
  }
}
