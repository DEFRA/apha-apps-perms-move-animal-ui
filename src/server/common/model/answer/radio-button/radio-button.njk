{% from "govuk/components/radios/macro.njk" import govukRadios %}
{{ govukRadios(assign({}, answer.viewModel, { errorMessage: errors[answer.config.payloadKey]})) }}
