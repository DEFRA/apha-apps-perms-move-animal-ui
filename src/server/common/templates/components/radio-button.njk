{% set items = [] %}
{% for key,option in answer.config.options %}
  {%
    set items = (items.push( {
      id: key,
      value: key,
      text: option.label
    }), items)
  %}
{% endfor %}

{{
  govukRadios({
    name: answer.config.payloadKey,
    id: answer.config.payloadKey,
    fieldset:{},
    value: value,
    items: items,
    errorMessage: errors[answer.config.payloadKey]
  })
}}
