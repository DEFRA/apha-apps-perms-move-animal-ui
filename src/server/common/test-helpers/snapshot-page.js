import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
