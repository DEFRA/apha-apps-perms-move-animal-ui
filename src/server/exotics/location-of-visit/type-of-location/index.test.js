import { describePageSnapshot } from '~/src/server/common/test-helpers/snapshot-page.js'
import { Answer, locationOfVisitPage } from './index.js'
import { RadioButtonAnswer } from '~/src/server/common/model/answer/radio-button/radio-button.js'
import { cphNumberPage } from '../cph-number/index.js'
import { hasACphNumberPage } from '../has-a-cph-number/index.js'
import { addressPage } from '../address/index.js'

const sectionKey = 'locationOfVisit'
const questionKey = 'typeOfLocation'
const pageUrl = '/exotics/location-of-visit/visit'
const page = locationOfVisitPage
const question = 'Where will the visit take place?'

const payload = {
  [questionKey]: 'some text'
}

describe('Answer', () => {
  it('should be a Radio button input', () => {
    expect(new Answer(payload)).toBeInstanceOf(RadioButtonAnswer)
  })

  it('should have the right payload key', () => {
    expect(Answer.config.payloadKey).toBe(questionKey)
  })

  it('should have validation for empty input', () => {
    expect(Answer.config.validation.empty).toBe(
      'Select where the visit will take place'
    )
  })

  it('should have the correct options', () => {
    expect(Answer.config.options.farm.label).toBe('Farm')
    expect(Answer.config.options['corporate-holding'].label).toBe(
      'Corporate holding'
    )
    expect(Answer.config.options['domestic-residence'].label).toBe(
      'Domestic residence'
    )
    expect(Answer.config.options.other.label).toBe(
      'Another location (such as a zoo or show)'
    )
  })
})

describe('LocationOfVisitPage', () => {
  it('should have the correct urlPath', () => {
    expect(page.urlPath).toBe(pageUrl)
  })

  it('should have the correct sectionKey', () => {
    expect(page.sectionKey).toBe(sectionKey)
  })

  it('should have the correct questionKey', () => {
    expect(page.questionKey).toBe(questionKey)
  })

  it('should have the correct question', () => {
    expect(page.question).toBe(question)
  })

  it('should have the correct Answer model', () => {
    expect(page.Answer).toBe(Answer)
  })

  describe('nextPage', () => {
    it.each([
      ['farm', cphNumberPage],
      ['corporate-holding', cphNumberPage],
      ['other', cphNumberPage]
    ])('for %s should return %s', (value, expectedPage) => {
      const answer = new Answer({ [questionKey]: value })
      const nextPage = page.nextPage(answer)
      expect(nextPage).toBe(expectedPage)
    })

    describe('domestic-residence', () => {
      const answer = new Answer({ [questionKey]: 'domestic-residence' })
      const liveAnimals = {
        whatIsMoving: 'live-animals'
      }

      it('should return cphNumberPage for pigs, sheep-and-goats, or cattle', () => {
        const state = { about: { ...liveAnimals, typeOfAnimal: 'pigs' } }
        const nextPage = page.nextPage(answer, state)
        expect(nextPage).toBe(hasACphNumberPage)
      })

      it('should return hasACphNumberPage for pigs, sheep-and-goats, or cattle', () => {
        const state = {
          about: { ...liveAnimals, typeOfAnimal: 'sheep-and-goats' }
        }
        const nextPage = page.nextPage(answer, state)
        expect(nextPage).toBe(hasACphNumberPage)
      })

      it('should return addressPage for other animals', () => {
        const state = { about: { ...liveAnimals, typeOfAnimal: 'dogs' } }
        const nextPage = page.nextPage(answer, state)
        expect(nextPage).toBe(addressPage)
      })
    })
  })

  describePageSnapshot({
    describes: 'content',
    it: 'should render expected response and content',
    pageUrl
  })
})
