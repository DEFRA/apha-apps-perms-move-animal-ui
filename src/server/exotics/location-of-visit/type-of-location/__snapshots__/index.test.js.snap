// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationOfVisitPage content should render expected response and content 1`] = `
"
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--l">
      <h1 class="govuk-fieldset__heading">
        Where will the visit take place?
      </h1>
    </legend>
    <div class="govuk-radios" data-module="govuk-radios">
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="typeOfLocation" name="typeOfLocation" type="radio" value="farm">
        <label class="govuk-label govuk-radios__label" for="typeOfLocation">
          Farm
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="corporate-holding" name="typeOfLocation" type="radio" value="corporate-holding">
        <label class="govuk-label govuk-radios__label" for="corporate-holding">
          Corporate holding
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="domestic-residence" name="typeOfLocation" type="radio" value="domestic-residence">
        <label class="govuk-label govuk-radios__label" for="domestic-residence">
          Domestic residence
        </label>
      </div>
      <div class="govuk-radios__item">
        <input class="govuk-radios__input" id="other" name="typeOfLocation" type="radio" value="other">
        <label class="govuk-label govuk-radios__label" for="other">
          Another location (such as a zoo or show)
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
